doc/boot_protocol.rst

```diff
@@ -0,0 +1,274 @@
+Security Kernel Loader
+######################
+
+The Security Kernel Loader (SKL) is an implementation of a Secure Loader for
+the SKINIT CPU instruction. The SKL loader takes a minimal approach relying on
+the DRTM Preamble, e.g. the system bootloader, to load into memory and
+configure the Security Kernel that the SKL will measure and execute.
+
+Memory Layout
+*************
+
+The DRTM Preamble is responsible for allocating a contiguous block of physical
+memory and will ensure it is marked as reserved in the memory map. The
+necessary size for the allocation will be found in the SKL header. The DRTM
+Preamble will place the SKL at the beginning of the allocation.
+
+Below is the layout of how the memory allocation is segmented::
+
+      +---     Security Kernel Loader    ---+
+      |                                     |
+      v                                     v
+      +--------+----------+-------------+-------------+--------------+
+      | Header |   Code   |  Boot Tags  | IOMMU Table |  Log Buffer  |
+      +--------+----------+-------------+-------------+--------------+
+      ^                                 ^                            ^
+      |                                 |                            |
+      +                                 +                            +
+  sklbuf_base                          60kb                   header.skl_size
+
+
+SKL Header
+==========
+
+The SKL header is the AMD Secure Loader Block (SLB) header extended to meet the
+needs of the SKL. The first two fields are mandated by the SLB and the SKL adds
+two additional fields which can be seen in the following representative struct,
+::
+
+    typedef struct __packed sl_header {
+        u16 entry_point;        /* Code Entry */
+        u16 boot_tags_offset;   /* RoT Measurement Length */
+        u16 info_table_offset;  /* Offset for the SKL Info Table */
+        u16 alloc_size;         /* Allocation Size */
+    } sl_header_t;
+
+SKL Allocation Size
+-------------------
+
+The length of Boot Tags are a runtime-determined size but cannot exceed the
+last possible start address for the IOMMU Table, that being 61440 byte offset
+from the beginning of the allocation. For build time calculation of the SKL
+allocation size, the IOMMU table will be considered to start at offset 61440.
+This renders the build time calculation to,
```

Comment:
The IOMMU isn't an implementation detail, and shouldn't be here.

The length of the boot tags is derived from the requirement to be outside of the measured region, but within the 64k DMA-protected region, and not overlapping with any other bits of that space that SKL is using.

Reply:
I get where you are going but this document is to tell boot loaders how then need to setup the loader and pass the necessary information. Since at least the first page of IOMMU and all of the boot params needs to be in the 64k page, they need to understand that when filling in the boot tags, they cannot exceed the last page.

Reply:
Is there anything that bootloader can do if there is not enough space for tags? If not, then there is no need to go into details. I think a build-time test in SKL for a minimal amount of space required for tags should suffice. We also may add another field to the header, but we definitely shouldn't hardcode it on the bootloader's side, since we may put other page-aligned data there, as mentioned in one of my previous comments.